Add flag to allow active descriptors in listeners

fredkilbourn opened the pull request:
In regards to issue #121:

There are times when listeners might want to make changes that trigger
further events in a collection.  I had a bug in my code around this
point and only realized that further collection changes are suppressed
by default.  This patch adds an override flag to allow for descriptors
to fire their listeners again even if currently active.

Signed-off-by: Fred Kilbourn <fred@fredk.com>

fredkilbourn commented:
Assuming you see validity in my suggestion, it is probably a good idea to add this flag to property-changes.js as well, but let me know.

kriskowal commented:
I have no philosophical objections to this change. When I factored observers into pop-observe, I omitted the feature entirely. If applied, it should be applied consistently across all observer classes.

marchant commented:
allowActive isn't a good make, it should be "allowsActive", but that's not really telling much, allowsNestedDispatch feels more appropriate, please update

fredkilbourn commented:
I've updated my branch as follows:
* Renamed allowsActive to allowsNestedDispatch
* Implemented in property-changes.js
* Cleaned up and unified comments + implementations across map-changes, property-changes, range-changes

If this pull does get accepted, will someone be updating the collections.js docs accordingly?

@kriskowal - As a relative outsider, I don't know if I fully understand the context of your comment, but if you plan on tearing out the isActive logic altogether will you just reject my pull and strip that logic out yourself?

kriskowal commented:
@fredkilbourn No, can’t. Backward compatibility across version 1 has to be maintained, and that’s the "latest" tag in npm.

fredkilbourn commented:
@kriskowal Okay, I get it.  So how's my updated pull request looking now?

fredkilbourn commented:
@stuk, @marchant, @kriskowal: Is there anything that needs to be fixed in this pull?  AFAIK it shouldn't have any backwards compatibility issues while adding better control over the nested triggering behavior.  Sorry to bump it's just been a number of months now.

Stuk commented:
@fredkilbourn sorry about that, you're right to give us a poke. I no longer have write access to this project, but I would suggest adding some tests for the new functionality so that it's clear that it works, and to ensure that it doesn't get broken in the future.

hthetiot commented:
Can you rebase on master @fredkilbourn ?

fredkilbourn commented:
@hthetiot Wow can't believe it's been 2 years!  I can fix it up but I'll have to set my environment back up and revisit the code, gimme a little time and I'll get to it.

hthetiot commented:
@francoisfrisch thanks there is a new test stack on master.

hthetiot commented:
you may want to check mergev2 branch that a work in progress that remove listen in favor of observe. it will be for v6.0.0 release. in the mean time feel free to rebase on master. but that may cause an issue for v6.

fredkilbourn commented:
@hthetiot Sincerest apologies, but the project I had this use case for is long defunct and I've really not been using your library for anything else lately, nor do I have the extra time to devote to this (however minor it is).

I'm just letting you know that I'm dropping off this thread and plan to delete my fork of your repo for now.  You're welcome to use my code / changes if you want them, but you'll have to finish it yourself.  Sorry!

PS - does anyone know if I delete my fork will it mess up this pull request?  I can leave it for a while if I need to / you want me to.
